openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/doc/man/man8/scls.8.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) scls.8.man,v 0.9.2.17 2006/04/05 09:43:50 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:43:50 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" scls.8.man,v
.\" Revision 0.9.2.17  2006/04/05 09:43:50  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database streams.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so streams.macros
.\"
.\"
.TH SCLS 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Administration"
.\"
.\"
.SH NAME
.B scls
\- produces a list of module and driver names
.\"
.\"
.SH SYNOPSIS
.PP
.HP 12
\fBscls\fR [\fIoptions\fR] [{\fB-c\fR|\fB--count\fR}|{\fB-l\fR|\fB--long\fR}] [\fImodule\fR ...]
.PD 0
.HP 12
\fBscls\fR {{\fB-h\fR|\fB--help\fR}|{\fB-V\fR|\fB--version\fR}|{\fB-C\fR|\fB--copying\fR}}
.PD
.\"
.\"
.SH OPTIONS
.PP
When issued without the
.BR -c " or " -l
options,
.B scls
lists the specified driver or module name if it is loaded on the
.I STREAMS
system, or, if no
.I module
name is specified, the names of all of the drivers and modules currently
loaded on the
.I STREAMS
system.
.TP
\fB-l\fR, \fB--long\fR [\fImodule\fR ...]
Produces a listing that shows the extension type, major number, and
information pertaining to the
.B module_info
structure, for each
.I module
name listed, or, if no
.I module
name is specified, all modules on the system.
.TP
\fB-c\fR, \fB--count\fR
Produces a listing showing the number of times an interface routine was
called for each
.I module
name listed, or, if no
.I module
name is specified, all modules on the system.  Information is obtained from
the
.BR module_stat (9)
structure if one is specified for the module.
.TP
\fB-q\fR, \fB--quiet\fR
Suppresses normal output.
.TP
\fB-v\fR, \fB--verbose\fR [\fIlevel\fR]
Increase or specify the verbosity level of the output.  This option may be
repeated.
.TP
\fB-h\fR, \fB--help\fR
When this options is encountered,
.B scls
displays usage information on
.I stdout
and exits.
.TP
\fB-V\fR, \fB--version\fR
When this option is encountered,
.B scls
displays versions information on
.I stdout
and exits.
.TP
\fB-C\fR, \fB--copying\fR
When this option is encountered,
.B scls
displays copying and copyright information on
.I stdout
and exits.
.PP
The
.BR -c " and " -l
options are mutually exclusive.
.\"
.\"
.SH AGRUMENTS
.PP
.TP
[\fImodule\fR \fI...\fR]
specifies the
.I STREAMS
modules names of interest, or it no
.I module
is specified, refers to all of the
.I STREAMS
modules loaded on the system.
.\"
.\"
.SH DESCRIPTION
.PP
The
.B scls
command provides a method for the caller to query the current
.I STREAMS
configuration.  Options can be used to produce enhanced lists.  Any further
parameters on the command line are module or driver names, and the output
produced is for only those names.
.\"
.\"
.SH FORMAT
.PP
The format of the output without the
.BR -c " or " -l
options is
.sp
.RB < name >,
where
.RS
.TP 12
.RB < name >
is the name of the driver or module.
.RE
.PP
The format of the output with the
.B -l
option is
.sp
.RB < type "> <" major "> <" mi_idnum "> <" mi_idname "> <" mi_minpsz "> <" mi_maxpsz "> <" mi_hiwat "> <" mi_lowat >,
where
.RS
.TP 12
.RB < type >
The module type.  This is either: \(lqmodule\(rq or \(lqdriver\(rq.
.PD 0
.TP
.RB < major >
The major device number
.RI ( %d )
when
.I type
is \(lqDriver\(rq, and a space otherwise.
.TP
.RB < mi_idnum >
The module ID number for the module
.RI ( %d ).
.TP
.RB < mi_idname >
The module name.
.TP
.RB < mi_minpsz >
The minimum packet size accepted.
.TP
.RB < mi_maxpsz >
The maximum packet size accepted.
.TP
.RB < mi_hiwat >
The (initial) high water mark for the queue pair.
.TP
.RB < mi_lowat >
The (initial) low water mark for the queue pair.
.PD
.RE
.PP
The format of the output with the
.B -c
option is
.sp
.RB < ms_pcnt "> <" ms_scnt "> <" ms_ocnt "> <" ms_acnt "> <" ms_flags >,
where
.RS
.TP 12
.RB < ms_pcnt >
The number of calls to the module's
.BR putp (9)
procedure, as a decimal number
.RI ( %lu ).
.PD 0
.TP
.RB < ms_scnt >
The number of calls to the module's
.BR srvp (9)
procedure, as a decimal number
.TP
.RB < ms_ocnt >
The number of calls to the module's
.BR qopen (9)
procedure.
.TP
.RB < ms_acnt >
The number of calls to the module's
.BR qadmin (9)
procedure.
.TP
.RB < ms_flags >
Boolean statistics flags.
.PD
.RE
.\"
.\"
.SH DIAGNOSTICS
.PP
An exit status of zero (0) indicates that the command was successful;
one (1) indicates that an error occurred and a diagnostic messages is 
printed to
.BR stderr ;
two (2) indicates that the option syntax was in error and a diagnostic message
is printed to
.BR stderr .
.PP
Errors may result from inadequate permissions, inability to open the
.BR nuls (4)
driver, inability to push the
.BR sc (4)
module.
.\"
.\"
.SH NOTICES
.PP
The
.B scls
command requires the
.BR sc (4)
.I STREAMS
module and the
.BR nuls (4)
.I STREAMS
driver.
If either one is not available, the
.B scls
command will not be successful.
.PP
Note that only the module information and module statistics associated with the write side of the
upper multiplexer will be displayed for a module or driver.  A module or driver that has more than
one
.BR module_info (9)
or
.BR module_stat (9)
structure in its definition, will only display the information associated with the
.I st_wrinit
member of the
.BR streamtab (9)
structure.
See
.BR sc (4),
.BR module_info (9),
.BR module_stat (9),
.BR streamtab (9)
for more information.
.\"
.\"
.SH EXAMPLES
.PP
Following is an example of the output of the
.B scls
command without options:
.RS
.nf
\fC\s-1
#> scls
clone
dlb
kinfo
log
nuls
echo
sad
pipe
xtisoUDP
xtisoTCP
ptm
pts
bba
lat
pppif
pppasync
pppcomp
bufcall
null
pass
errm
ptem
spass
rspass
pipemod
timod
tirdwr
ldtty
ifnet
\s+1\fP
.fi
.RE
.PP
Following is an example of the output of the
.B scls
command with the
.B -l
option:
.RS
.nf
\fC\s-1
#> scls -l
clone           driver     32       0 xxxx xxxxx xxxxx xxxxx
dlb             driver     52    5010 xxxx xxxxx xxxxx xxxxx
kinfo           driver     53    5020 xxxx xxxxx xxxxx xxxxx
log             driver     54      44 xxxx xxxxx xxxxx xxxxx
nuls            driver     55    5001 xxxx xxxxx xxxxx xxxxx
echo            driver     56    5000 xxxx xxxxx xxxxx xxxxx
sad             driver     57      45 xxxx xxxxx xxxxx xxxxx
pipe            driver     58    5304 xxxx xxxxx xxxxx xxxxx
xtisoUDP        driver     59    5010 xxxx xxxxx xxxxx xxxxx
xtisoTCP        driver     60    5010 xxxx xxxxx xxxxx xxxxx
ptm             driver     63    7609 xxxx xxxxx xxxxx xxxxx
pts             driver      6    7608 xxxx xxxxx xxxxx xxxxx
bba             driver     64   24880 xxxx xxxxx xxxxx xxxxx
lat             driver      5       5 xxxx xxxxx xxxxx xxxxx
pppif           module           6002 xxxx xxxxx xxxxx xxxxx
pppasync        module           6000 xxxx xxxxx xxxxx xxxxx
pppcomp         module           6001 xxxx xxxxx xxxxx xxxxx
bufcall         module              0 xxxx xxxxx xxxxx xxxxx
null            module           5002 xxxx xxxxx xxxxx xxxxx
pass            module           5003 xxxx xxxxx xxxxx xxxxx
errm            module           5003 xxxx xxxxx xxxxx xxxxx
ptem            module           5003 xxxx xxxxx xxxxx xxxxx
spass           module           5007 xxxx xxxxx xxxxx xxxxx
rpass           module           5008 xxxx xxxxx xxxxx xxxxx
pipemod         module           5303 xxxx xxxxx xxxxx xxxxx
timod           module           5006 xxxx xxxxx xxxxx xxxxx
tirdwr          module              0 xxxx xxxxx xxxxx xxxxx
ldtty           module           7701 xxxx xxxxx xxxxx xxxxx
ifnet           module           5501 xxxx xxxxx xxxxx xxxxx
\s+1\fP
.fi
.RE
.PP
Following is an example of the output of the
.B scls
command with the
.B -c
option:
.RS
.nf
\fC\s-1
#> scls -c
clone           %lu     %lu     %lu     %lu     %lx
dlb             %lu     %lu     %lu     %lu     %lx
kinfo           %lu     %lu     %lu     %lu     %lx
log             %lu     %lu     %lu     %lu     %lx
nuls            %lu     %lu     %lu     %lu     %lx
echo            %lu     %lu     %lu     %lu     %lx
sad             %lu     %lu     %lu     %lu     %lx
pipe            %lu     %lu     %lu     %lu     %lx
xtisoUDP        %lu     %lu     %lu     %lu     %lx
xtisoTCP        %lu     %lu     %lu     %lu     %lx
ptm             %lu     %lu     %lu     %lu     %lx
pts             %lu     %lu     %lu     %lu     %lx
bba             %lu     %lu     %lu     %lu     %lx
lat             %lu     %lu     %lu     %lu     %lx
pppif           %lu     %lu     %lu     %lu     %lx
pppasync        %lu     %lu     %lu     %lu     %lx
pppcomp         %lu     %lu     %lu     %lu     %lx
bufcall         %lu     %lu     %lu     %lu     %lx
null            %lu     %lu     %lu     %lu     %lx
pass            %lu     %lu     %lu     %lu     %lx
errm            %lu     %lu     %lu     %lu     %lx
ptem            %lu     %lu     %lu     %lu     %lx
spass           %lu     %lu     %lu     %lu     %lx
rpass           %lu     %lu     %lu     %lu     %lx
pipemod         %lu     %lu     %lu     %lu     %lx
timod           %lu     %lu     %lu     %lu     %lx
tirdwr          %lu     %lu     %lu     %lu     %lx
ldtty           %lu     %lu     %lu     %lu     %lx
ifnet           %lu     %lu     %lu     %lu     %lx
\s+1\fP
.fi
.RE
.\"
.\"
.SH DEVICES
.PP
.TP
.BR sc (4)
.I STREAMS
configuration module.
.TP
.B /dev/nuls
.I STREAMS
null device, see
.BR nuls (4).
.\"
.\"
.SH FILES
.PP
.TP
.B /usr/sbin/scls
contains the
.B scls
command.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR sc (4),
.BR nuls (4),
.BR strinfo (8),
.BR strload (8),
.BR strsetup (8).
.\"
.\"
.SH BUGS
.PP
.B scls
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.IR AIX \(rg.
This is an
.IR AIX \(rg-specific
command.
.B scls
is provided in
.I @PACKAGE_TITLE@
for compatibility with
.IR AIX \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IR AIX \(rg
does not document long options.
.IP \(em
.IR AIX \(rg
does not document the
.BR -q ", " -v ", " -h ", " -V " or " -C
options.
.IP \(em
Options
.BR --help " and " --version
are provided for compatibility with GNU coding standards;
.BR --copying ,
OpenSS7 coding standards.
.IP \(em
The format of the output from the
.B scls
command with the
.BR -c " or " -l
options may be different from that of
.IR AIX \(rg.
.\"
.\"
.SH CONFORMANCE
.PP
.IR AIX \(rg
documentation.
.[
aixdoc
.]
.\"
.\"
.SH HISTORY
.PP
.B scls
first appeared in
.IR AIX \(rg.
.\"
.\"
.[
$LIST$
.]
.TI


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/src/util/scls.c
/*****************************************************************************

 @(#) scls.c,v OpenSS7-0_9_2_D_rc2(0.9.2.20) 2006/03/03 10:57:17

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; version 2 of the License.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/03/03 10:57:17 by brian

 -----------------------------------------------------------------------------

 scls.c,v
 Revision 0.9.2.20  2006/03/03 10:57:17  brian
 - 32-bit compatibility support, updates for release

 Revision 0.9.2.19  2006/02/20 10:59:30  brian
 - updated copyright headers on changed files

 *****************************************************************************/

#ident "@(#) scls.c,v OpenSS7-0_9_2_D_rc2(0.9.2.20) 2006/03/03 10:57:17"

static char const ident[] =
    "scls.c,v OpenSS7-0_9_2_D_rc2(0.9.2.20) 2006/03/03 10:57:17";

/* 
 *  AIX Utility: scls - Produces a list of module and driver names.
 */

#define _XOPEN_SOURCE 600

#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/ioctl.h>

#ifdef _GNU_SOURCE
#include <getopt.h>
#endif

#include <stropts.h>
#include <sys/sc.h>

static int debug = 0;			/* default no debug */
static int output = 1;			/* default normal output */

static void
version(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stdout, "\
%2$s\n\
Copyright (c) 2001-2006  OpenSS7 Corporation.  All Rights Reserved.\n\
Distributed under GPL Version 2, included here by reference.\n\
See `%1$s --copying' for copying permissions.\n\
", argv[0], ident);
}

static void
usage(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stderr, "\
Usage:\n\
    %1$s [options] [{-c|--count}|{-l|--long}] [MODULE ...]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
", argv[0]);
}

static void
help(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stdout, "\
Usage:\n\
    %1$s [options] [{-c|--count}|{-l|--long}] [MODULE ...]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
Arguments:\n\
    MODULE ...\n\
	specific drivers and modules to list instead of all drivers and\n\
	modules\n\
Options:\n\
    -c, --count\n\
        print module_stats only\n\
    -l, --long\n\
        print module_info and module_stats\n\
    -q, --quiet\n\
        suppress output\n\
    -d, --debug [LEVEL]\n\
        increase or set debugging verbosity\n\
    -v, --verbose [LEVEL]\n\
        increase or set output verbosity\n\
    -h, --help\n\
        prints this usage information and exits\n\
    -V, --version\n\
        prints the version and exits\n\
    -C, --copying\n\
        prints copying permissions and exits\n\
", argv[0]);
}

static void
copying(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	fprintf(stdout, "\
--------------------------------------------------------------------------------\n\
%1$s\n\
--------------------------------------------------------------------------------\n\
Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com>\n\
Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>\n\
\n\
All Rights Reserved.\n\
--------------------------------------------------------------------------------\n\
This program is free software; you can  redistribute  it and/or modify  it under\n\
the terms  of the GNU General Public License  as  published by the Free Software\n\
Foundation; version  2  of  the  License.\n\
\n\
This program is distributed in the hope that it will  be useful, but WITHOUT ANY\n\
WARRANTY; without even  the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU General Public License for more details.\n\
\n\
You should  have received a copy of the GNU  General  Public License  along with\n\
this program; if not, write to the Free Software Foundation, Inc., 675 Mass Ave,\n\
Cambridge, MA 02139, USA.\n\
--------------------------------------------------------------------------------\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the U.S. Government (\"Government\"), the following provisions apply to you. If\n\
the Software is supplied by the  Department of Defense (\"DoD\"), it is classified\n\
as \"Commercial  Computer  Software\"  under  paragraph  252.227-7014  of the  DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the  license rights granted\n\
herein (the license rights customarily provided to non-Government users). If the\n\
Software is supplied to any unit or agency of the Government  other than DoD, it\n\
is  classified as  \"Restricted Computer Software\" and the Government's rights in\n\
the Software  are defined  in  paragraph 52.227-19  of the  Federal  Acquisition\n\
Regulations (\"FAR\")  (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of  the  NASA  Supplement  to the FAR (or any  successor\n\
regulations).\n\
--------------------------------------------------------------------------------\n\
Commercial  licensing  and  support of this  software is  available from OpenSS7\n\
Corporation at a fee.  See http://www.openss7.com/\n\
--------------------------------------------------------------------------------\n\
", ident);
}

enum { CMN_NONE, CMN_NAMES, CMN_LONG, CMN_COUNT, } command = CMN_NONE;

void
printit(struct sc_mlist *l, int cmd)
{
	if (output <= 0 || l->major == -1)
		return;
	fprintf(stdout, "%s", l->name);
	switch (cmd) {
	case CMN_LONG:
		if (l->major != 0) {
			fprintf(stdout, "\tdevice");
			fprintf(stdout, "\t%ld", (long) l->major);
		} else {
			fprintf(stdout, "\tmodule");
			fprintf(stdout, "\t-");
		}
		fprintf(stdout, "\t%u", l->mi.mi_idnum);
		fprintf(stdout, "\t%ld", (long) l->mi.mi_minpsz);
		fprintf(stdout, "\t%ld", (long) l->mi.mi_maxpsz);
		fprintf(stdout, "\t%ld", (long) l->mi.mi_hiwat);
		fprintf(stdout, "\t%ld", (long) l->mi.mi_lowat);
	case CMN_COUNT:
		fprintf(stdout, "\t%ld", (long) l->ms.ms_pcnt);
		fprintf(stdout, "\t%ld", (long) l->ms.ms_scnt);
		fprintf(stdout, "\t%ld", (long) l->ms.ms_ocnt);
		fprintf(stdout, "\t%ld", (long) l->ms.ms_acnt);
		fprintf(stdout, "\t%x", l->ms.ms_flags);
	}
	fprintf(stdout, "\n");
};

int
main(int argc, char *argv[])
{
	int i, fd, count;
	struct sc_list *list;

	for (;;) {
		int c, val;

#if defined _GNU_SOURCE
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{"long",	no_argument,		NULL, 'l'},
			{"count",	no_argument,		NULL, 'c'},
			{"quiet",	no_argument,		NULL, 'q'},
			{"debug",	optional_argument,	NULL, 'D'},
			{"verbose",	optional_argument,	NULL, 'v'},
			{"help",	no_argument,		NULL, 'h'},
			{"version",	no_argument,		NULL, 'V'},
			{"copying",	no_argument,		NULL, 'C'},
			{"?",		no_argument,		NULL, 'H'},
			{ 0, }
		};
		/* *INDENT-ON* */

		c = getopt_long_only(argc, argv, "lcqD::v::hVC?W:", long_options, &option_index);
#else				/* defined _GNU_SOURCE */
		c = getopt(argc, argv, "lcqDvhVC?");
#endif				/* defined _GNU_SOURCE */
		if (c == -1) {
			if (debug)
				fprintf(stderr, "%s: done options processing\n", argv[0]);
			break;
		}
		switch (c) {
		case 0:
			goto bad_usage;
		case 'l':
			if (command != CMN_NONE)
				goto bad_option;
			if (debug)
				fprintf(stderr, "%s: setting long command\n", argv[0]);
			command = CMN_LONG;
			break;
		case 'c':
			if (command != CMN_NONE)
				goto bad_option;
			if (debug)
				fprintf(stderr, "%s: setting count command\n", argv[0]);
			command = CMN_COUNT;
			break;
		case 'D':	/* -D, --debug */
			if (debug)
				fprintf(stderr, "%s: increasing debug verbosity\n", argv[0]);
			if (optarg == NULL) {
				debug++;
			} else {
				if ((val = strtol(optarg, NULL, 0)) < 0)
					goto bad_option;
				debug = val;
			}
			break;
		case 'q':	/* -q, --quiet */
			if (debug)
				fprintf(stderr, "%s: suppressing normal output\n", argv[0]);
			debug = 0;
			output = 0;
			break;
		case 'v':	/* -v, --verbose [level] */
			if (debug)
				fprintf(stderr, "%s: increasing output verbosity\n", argv[0]);
			if (optarg == NULL) {
				output++;
				break;
			}
			if ((val = strtol(optarg, NULL, 0)) < 0)
				goto bad_option;
			output = val;
			break;
		case 'h':	/* -h, --help */
		case 'H':	/* -H, --? */
			if (debug)
				fprintf(stderr, "%s: printing help message\n", argv[0]);
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			if (debug)
				fprintf(stderr, "%s: printing version message\n", argv[0]);
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			if (debug)
				fprintf(stderr, "%s: printing copying message\n", argv[0]);
			copying(argc, argv);
			exit(0);
		case '?':
		default:
		      bad_option:
			optind--;
			goto bad_nonopt;
		      bad_nonopt:
			if (output || debug) {
				if (optind < argc) {
					fprintf(stderr, "%s: syntax error near '", argv[0]);
					while (optind < argc)
						fprintf(stderr, "%s ", argv[optind++]);
					fprintf(stderr, "'\n");
				} else {
					fprintf(stderr, "%s: missing option or argument", argv[0]);
					fprintf(stderr, "\n");
				}
				fflush(stderr);
			      bad_usage:
				usage(argc, argv);
			}
			exit(2);
		}
	}
	if (debug)
		fprintf(stderr, "%s: opening /dev/nuls\n", argv[0]);
	if ((fd = open("/dev/nuls", O_RDWR)) < 0) {
		if (debug)
			fprintf(stderr, "%s: could not open /dev/nuls\n", argv[0]);
		perror(argv[0]);
		exit(1);
	}
	if (debug)
		fprintf(stderr, "%s: pushing sc module\n", argv[0]);
	if (ioctl(fd, I_PUSH, "sc") < 0) {
		if (debug)
			fprintf(stderr, "%s: could not push sc module\n", argv[0]);
		perror(argv[0]);
		exit(1);
	}
	if (debug)
		fprintf(stderr, "%s: getting size of list\n", argv[0]);
	/* go out and get all the names anyway */
	if ((count = ioctl(fd, SC_IOC_LIST, NULL)) < 0) {
		if (debug)
			fprintf(stderr, "%s: could not perform SC_IOC_LIST command\n", argv[0]);
		perror(argv[0]);
		exit(1);
	}
	if (debug)
		fprintf(stderr, "%s: size of list = %d\n", argv[0], count);
	/* get entire list */
	if ((list = malloc(sizeof(struct sc_list) + count * sizeof(struct sc_mlist))) == NULL) {
		if (debug)
			fprintf(stderr, "%s: could not allocate memory\n", argv[0]);
		fprintf(stderr, "%s: %s\n", argv[0], strerror(ENOMEM));
		fflush(stderr);
		exit(1);
	}
	list->sc_nmods = count;
	list->sc_mlist = (struct sc_mlist *) (list + 1);
	if (ioctl(fd, SC_IOC_LIST, list) < 0) {
		if (debug)
			fprintf(stderr, "%s: could not perform second SC_IOC_LIST command\n", argv[0]);
		perror(argv[0]);
		exit(1);
	}
	if (optind < argc) {
		int ret = 0;

		/* have module name arguments - iterate through them */
		for (; optind < argc; optind++) {
			for (i = 0; i < count; i++)
				if (strncmp(argv[optind], list->sc_mlist[i].name, FMNAMESZ)
				    == 0)
					break;
			if (i < count)
				printit(&list->sc_mlist[i], command);
			else {
				fprintf(stderr, "%s: %s: %s\n", argv[0], argv[optind],
					strerror(ENXIO));
				fflush(stderr);
				ret = 1;
			}
		}
		exit(ret);
	} else {
		/* have no module name arguments - operate on all */
		for (i = 0; i < count; i++)
			printit(&list->sc_mlist[i], command);
	}
	if (ioctl(fd, I_POP, NULL) < 0) {
		perror(argv[0]);
	}
	if (close(fd) < 0) {
		perror(argv[0]);
	}
	exit(0);
}
